### ISSUE TITLE
Promise wrapper on createConnection is problematic in top-level code

### ISSUE THREAD

**Issue by jdhiro** [sidorares/node-mysql2#466]
The non-promise version of createConnection is not asynchronous. Unlike the database calls, no callback is required here.

`var connection = mysql.createConnection({host:'localhost', user: 'root', database: 'test'});`

However, the promise wrapper version comes back as a promise.

`let connection = await mysql.createConnection({host:'localhost', user: 'root', database: 'test'});`

This wouldn't be a problem, except that Node/V8 don't currently allow await outside of an async function, including at the top level of an app. This means that the example code doesn't work when you drop it into a simple Node app. The easy solution that I've come up for this, is to wrap createConnection in an IIFE.

```let connection

(async function() {
  try {
    connection = await mysql.createConnection(dbconfig)
  } catch(err){
    console.error(err)
  }
})()```

This seems unnecessarily complex, since this isn't a method that needs to be async. I would propose that createConnection remains a synchronous function, the same way it is in the non-promise version.

**Comment by sidorares** [comment]
hi @jdhiro 

I see your point, I wish 7+ node had repl and top level module wrapper async functions

Here are my thoughts: the only correct way of using connection, even in simple script is like this:

```js
   const conn = mysql.createConnection(opts);
   conn.on('error', (err) => { /* 
      handle non-query errors: connection time, network etc
      connection time errors usually can be handled without 
      dedicated conn.on('error') handler because usually there 
      is a command in the queue and fatal errors are reported 
      down to all commands. However, when there is no 
      command in the queue, or command with no callback
      and there is connection error, error must be handled by
      conn error handler, otherwise it'll be re-thrown and 
      crash process
     */ })
    conn.query('do some sql', (err, res) => {
      if (err) {
         return err;
      }
    })
```

async `.createConnection()` helps a little with this, in a way that if there is connection error and no command error can be caught. Unfortunately it's still possible for 'error' connection event to crash process so you must attach listener. My advise is to always use pool, even in scripts - pool handles this error logic automatically.

I'll think about changing `.createConnection()` to be sync. Maybe it's possible to make this backwards compatible

as a work around - if you prefer async wrappers but want to have `.createConnection()` in top lewel of script without extra wrapping - just don't use await:

```js
const myAsyncFunc = async (conn, x) =>  {
   const [rows, fields] = conn.query('select 1+?', x)
   console.log(rows);
}

const mysql = require('mysql2/promise');
mysql.createConnection({}).then( (conn) => {
   myAsyncFunc(conn, 1);
   myAsyncFunc(conn, 2);
}).catch((e) => {
  console.log('error!!!', e);
})
```

**Comment by kswope** [comment]
This is a show stopper for me.  I'm not sure why createConnection had to be async, it doesn't make any sense.

**Comment by sidorares** [comment]
@kswope can you use `createPool` instead? It's sync

The reason `createConnection` is async is to allow you to catch connection time runtime errors:

```js
   let conn;
   try {
      const credentials = await getCredentials();
      conn = await mysql.createConnection(credentials)
   } catch(e) {
      // fallback or cleanup 
   }
```

**Comment by kswope** [comment]
> The reason createConnection is async is to allow you to catch connection time runtime errors:

You can still get catch connection time runtime errors with a sync method as well, I don't see your point.

knex has no problem doing this (even using this module), and I could promisify the other mysql drivers query method and get the same thing.

**Comment by kswope** [comment]
Ok I see, you want to handle the situation where the connection is dependent on previous network call.  Don't think I've ever seen that before, must be a rare situation.  Maybe a createConnectionSync would be a solution.

**Comment by sidorares** [comment]
currently connection object tries to perform network connection and handshake immediately. For a sync `createConnection()` you would have to handle error on connection object

```js
   const conn = mysql.createConnectionSync (params);
   conn.on('error', err => {
     // handle network or handshake errors here
     // network: host/port not reachable
     // Handshake: incorrect password, incorrect access (ip, table restrictions etc), too many users
   })
   const data = await getData();
   const result = conn.execute('select foo from bar where id = ?', [data.id]);
``` 
   
Is there any reason `createPool` won't work for you where you'd use `createConnectionSync` ?

**Comment by kswope** [comment]
Ok I'm just being dumb.  I assumed createConnection was for pooling (maybe because I'm also using knex and it makes pools be default), and there's no mention of pooling in the docs ( I know its covered by the other mysql module ), and then I assumed that createPool didn't work with 'mysql2/promise' because the example doesn't cover it.

**Comment by sidorares** [comment]
Yeah, maybe it should have been default to create pool. Currently createConnection creates single connection. I always recommend pool even if it's not required for performance reason, because pool adds useful error logic

**Comment by kswope** [comment]
If you are trying to stick with the behavior of mysqljs/mysql then you've made the right choice in my opinion.

>  MySQL2 team is working together with mysqljs/mysql team to factor out shared code and move it under mysqljs organisation.

Is that still happening?

**Comment by dougwilson** [comment]
Yea, in fact was just pulling out the AWS certs to share right now :waves:
